File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

unordered_map<char, int> un_m;

bool isDelete(string today, string dday) {
    int today_year = stoi(today.substr(0, 4));
    int dday_year = stoi(dday.substr(0,4));
    
    int today_month = stoi(today.substr(5,2));
    int dday_month = stoi(dday.substr(5,2));
    
    int today_date = stoi(today.substr(8,2));
    int dday_date = stoi(dday.substr(8,2));
    
    if(today_year == dday_year) {
        if(today_month == dday_month) {
            if(today_date >= dday_date) return true;
            else return false;
        }
        else if(today_month > dday_month)
            return true;
        else
            return false;
    }
    else if(today_year > dday_year)
        return true;
    else
        return false;
}

vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    
    for(string s : terms) {
        char c = s[0];
        
        int month;
        string tmp;
        tmp.append(s.substr(2));
        
        un_m[c] = stoi(tmp);
    }
    
    int idx = 1;
    for(string s : privacies) {
        int year = stoi(s.substr(0, 4));
        int month = stoi(s.substr(5,2));
        
        int term = un_m[s[11]];
        
        month += term;
        
        if(month > 12) {
            int cnt = month / 12;
            month %= 12;
            year += cnt;
            
            if(month == 0)
                month = 12, year--;
        }
        
        string newMonth;
        if(month < 10)
            newMonth.push_back('0');
        
        newMonth.append(to_string(month));
        
        s.replace(0,4,to_string(year));
        s.replace(5,2,newMonth);
        
        if(isDelete(today, s.substr(0,10)))
            answer.push_back(idx);
        
        idx++;
    }
    
    return answer;
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {
    
    public int[] solution(String today, String[] terms, String[] privacies) {
        List<Integer> answer = new ArrayList<>();
        
        Date todayDate = new Date(today);
        int idx = 1;
        for(String p : privacies) {
            Date date = new Date(p);
            char term = p.charAt(11);
            
            if(isExpiration(date, todayDate, convertToIntTerm(term, terms))) {
                answer.add(idx);
            }
            
            idx++;
        }
        
        
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
    
    private boolean isExpiration(Date target, Date today, int term) {
        target.month += term;
        while(target.month > 12) {
            target.month -= 12;
            target.year += 1;
        }
        
        System.out.println("[Expiration Date] => " + target.year + "." + target.month + "." + target.day);
        
        if(target.year < today.year)
            return true;
        
        if(target.year == today.year)
            if(target.month < today.month)
                return true;
        
        if(target.year == today.year)
            if(target.month == today.month)
                if(target.day <= today.day)
                    return true;
        
        return false;
    }
    
    private int convertToIntTerm(char t, String[] terms) {
        for(String term : terms) {
            char c = term.charAt(0);
            int i = Integer.parseInt(term.substring(2));
            
            if(c == t) return i;
        }
        return -1;
    }
}

class Date {
    int year;
    int month;
    int day;
    
    Date(String date) {
        this.year = Integer.parseInt(date.substring(0,4));
        this.month = Integer.parseInt(date.substring(5,7));
        this.day = Integer.parseInt(date.substring(8,10));
    }
}